lib/chef/knife/cloud/google_service.rb

```diff
@@ -0,0 +1,503 @@
+        [
+          "https://www.googleapis.com/auth/cloud-platform",
+          "https://www.googleapis.com/auth/compute",
+        ]
+      )
+    end
+
+    def create_server(options = {})
+      validate_server_create_options!(options)
+
+      ui.msg("Creating instance...")
+
+      instance_object = instance_object_for(options)
+      wait_for_operation(connection.insert_instance(project, zone, instance_object))
+      wait_for_status("RUNNING") { get_server(options[:name]) }
+
+      ui.msg("Instance created!")
+
+      get_server(options[:name])
+    end
+
+    def delete_server(instance_name)
+      begin
+        instance = get_server(instance_name)
+      rescue Google::Apis::ClientError
+        ui.warn("Unable to locate instance #{instance_name} in project #{project}, zone #{zone}")
+        return
+      end
+
+      server_summary(instance)
+      ui.confirm("Do you really want to delete this instance")
+
+      ui.msg("Deleting instance #{instance_name}...")
+
+      wait_for_operation(connection.delete_instance(project, zone, instance_name))
+
+      ui.msg("Instance #{instance_name} deleted successfully.")
+    end
+
+    def get_server(instance_name)
+      connection.get_instance(project, zone, instance_name)
+    end
+
+    def create_disk(name, size, type, source_image = nil)
+      disk = Google::Apis::ComputeV1::Disk.new
+      disk.name    = name
+      disk.size_gb = size
+      disk.type    = disk_type_url_for(type)
+
+      ui.msg("Creating a #{size} GB disk named #{name}...")
+
+      wait_for_operation(connection.insert_disk(project, zone, disk, source_image: source_image))
+
+      ui.msg("Waiting for disk to be ready...")
+
+      wait_for_status("READY") { connection.get_disk(project, zone, name) }
+
+      ui.msg("Disk created successfully.")
+    end
+
+    def delete_disk(name)
+      begin
+        connection.get_disk(project, zone, name)
+      rescue Google::Apis::ClientError
+        ui.warn("Unable to locate disk #{name} in project #{project}, zone #{zone}")
+        return
+      end
+
+      ui.confirm("Do you really want to delete disk #{name}")
+
+      ui.msg("Deleting disk #{name}...")
+      wait_for_operation(connection.delete_disk(project, zone, name))
+      ui.msg("Disk #{name} deleted successfully.")
+    end
+
+    def list_servers
+      instances = paginated_results(:list_instances, :items, project, zone)
+      return [] if instances.nil?
+
+      instances.each_with_object([]) do |instance, memo|
+        memo << OpenStruct.new(
+          name:         instance.name,
+          status:       instance.status,
+          machine_type: machine_type_for(instance),
+          network:      network_for(instance),
+          private_ip:   private_ip_for(instance),
+          public_ip:    public_ip_for(instance)
+        )
+      end
+    end
+
+    def list_zones
+      zones = paginated_results(:list_zones, :items, project)
+      return [] if zones.nil?
+
+      zones
```

Comment:
Yeah, I can go back and change that. Originally, I thought some of these responses would also return an empty string based on initial docs, so there was a `|| blah.empty?` there too. That's not actually the case, so I can clean this up.